Improve the tracing quickstart guide for LS (#522)
langchain-ai/docs: src/langsmith/observability-quickstart.mdx

export LANGSMITH_WORKSPACE_ID="<your-workspace-id>"
```

> The example app in this quickstart will use OpenAI as the LLM provider. You can adapt the example for your app's LLM provider.

Author: Done!

async function rag(question: string) {
  const docs = retriever(question);
  const systemMessage =
    "Answer the user's question using only the provided information below:\n" +
    docs.join("\n");

  // This call is not traced yet
  const resp = await client.chat.completions.create({
    model: "gpt-4o-mini",
    messages: [
      { role: "system", content: systemMessage },
      { role: "user", content: question },
    ],
  });

  return resp.choices[0].message?.content;
}

(async () => {
  console.log(await rag("Where did Harrison work?"));
})();
```

</CodeGroup>

## 4. Trace LLM calls

To start, you’ll trace all your OpenAI calls. LangSmith provides wrappers:

- Python: [`wrap_openai`](https://docs.smith.langchain.com/reference/python/wrappers/langsmith.wrappers._openai.wrap_openai)
- TypeScript: [`wrapOpenAI`](https://docs.smith.langchain.com/reference/js/functions/wrappers_openai.wrapOpenAI)

This snippet wraps the OpenAI client so that every subsequent model call is logged automatically as a traced child run in LangSmith.

1. Include the highlighted lines in your app file:

    <CodeGroup>

    ```python Python highlight={2,7}
    from openai import OpenAI

>     from langsmith.wrappers import wrap_openai  # traces SDK calls

Reviewer: ```suggestion
    from langsmith.wrappers import wrap_openai  # traces openai calls
```

    def retriever(query: str):
        return ["Harrison worked at Kensho"]

>     client = wrap_openai(OpenAI())  # child run for the model call

Reviewer: ```suggestion
    client = wrap_openai(OpenAI())  # log traces by wrapping the model calls 
```

    def rag(question: str) -> str:
        docs = retriever(question)
        system_message = (
            "Answer the user's question using only the provided information below:\n"
            + "\n".join(docs)
        )
        resp = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {"role": "system", "content": system_message},
                {"role": "user", "content": question},
            ],
        )
        return resp.choices[0].message.content

    if __name__ == "__main__":
        print(rag("Where did Harrison work?"))
    ```

    ```typescript TypeScript highlight={3,9}
    import "dotenv/config";
    import OpenAI from "openai";

>     import { wrapOpenAI } from "langsmith/wrappers"; // traces SDK calls

Reviewer: ```suggestion
    import { wrapOpenAI } from "langsmith/wrappers"; // traces openai calls
```

    function retriever(query: string): string[] {
    return ["Harrison worked at Kensho"];
    }

>     const client = wrapOpenAI(new OpenAI()); // child run for the model call

Reviewer: ```suggestion
    const client = wrapOpenAI(new OpenAI()); // log traces by wrapping the model calls 
```

    async function rag(question: string) {
    const docs = retriever(question);
    const systemMessage =
        "Answer the user's question using only the provided information below:\n" +
        docs.join("\n");

    const resp = await client.chat.completions.create({
        model: "gpt-4o-mini",
        messages: [
        { role: "system", content: systemMessage },
        { role: "user", content: question },
        ],
    });

    return resp.choices[0].message?.content;
    }

    (async () => {
    console.log(await rag("Where did Harrison work?"));
    })();
    ```

    </CodeGroup>

1. Call your application:

    <CodeGroup>

    ```bash Python
    python app.py
    ```

    ```bash TypeScript
    npx ts-node app.ts
    ```

    </CodeGroup>

    You'll receive the following output:

    ```
    Harrison worked at Kensho.
    ```

> 1. In the [platform UI](smith.langchain.com), select the **default** workspace (or the workspace you specified in [Step 2](#2-set-up-environment-variables)). Select the tracing project and then the run under the **Runs** tab. You'll find a trace of just the OpenAI call.

Reviewer: ```suggestion
1. In the [LangSmith UI](https://smith.langchain.com), navigate to the **default** Tracing Project for your workspace (or the workspace you specified in [Step 2](#2-set-up-environment-variables). You'll see the OpenAI call we just instrumented.
```

<div style={{ textAlign: 'center' }}>
<img
    className="block dark:hidden"
    src="/langsmith/images/trace-quickstart-llm-call.png"
    alt="LangSmith UI showing an LLM call trace called ChatOpenAI with a system and human input followed by an AI Output."
/>

<img
    className="hidden dark:block"
    src="/langsmith/images/trace-quickstart-llm-call-dark.png"
    alt="LangSmith UI showing an LLM call trace called ChatOpenAI with a system and human input followed by an AI Output."
/>
</div>

## 5. Trace an entire application

You can also use the `traceable` decorator for [Python](https://docs.smith.langchain.com/reference/python/run_helpers/langsmith.run_helpers.traceable) or [TypeScript](https://langsmith-docs-bdk0fivr6-langchain.vercel.app/reference/js/functions/traceable.traceable) to trace your entire application instead of just the LLM calls.

1. Include the highlighted code in your app file:

    <CodeGroup>

    ```python Python highlight={3,10}
    from openai import OpenAI
    from langsmith.wrappers import wrap_openai

>     from langsmith import traceable  # parent run for your function

Reviewer: ```suggestion
    from langsmith import traceable  
```

    def retriever(query: str):
        return ["Harrison worked at Kensho"]

>     client = wrap_openai(OpenAI())  # keep this so the LLM call nests as a child run

Reviewer: ```suggestion
    client = wrap_openai(OpenAI())  # keep this to capture the prompt and response from the LLM
```

    @traceable
    def rag(question: str) -> str:
        docs = retriever(question)
        system_message = (
            "Answer the user's question using only the provided information below:\n"
            + "\n".join(docs)
        )
        resp = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {"role": "system", "content": system_message},
                {"role": "user", "content": question},
            ],
        )
        return resp.choices[0].message.content

    if __name__ == "__main__":
        print(rag("Where did Harrison work?"))
    ```

    ```typescript TypeScript highlight={3,11}
    import "dotenv/config";
    import OpenAI from "openai";

>     import { wrapOpenAI, traceable } from "langsmith/wrappers"; // parent + child

Reviewer: ```suggestion
    import { wrapOpenAI, traceable } from "langsmith/wrappers"; 
```

    function retriever(query: string): string[] {
    return ["Harrison worked at Kensho"];
    }

>     const client = wrapOpenAI(new OpenAI()); // keep this to nest the LLM call

Reviewer: ```suggestion
    const client = wrapOpenAI(new OpenAI()); // # keep this to capture the prompt and response from the LLM
```

    const rag = traceable(async (question: string) => {
    const docs = retriever(question);
    const systemMessage =
        "Answer the user's question using only the provided information below:\n" +
        docs.join("\n");

    const resp = await client.chat.completions.create({
        model: "gpt-4o-mini",
        messages: [
        { role: "system", content: systemMessage },
        { role: "user", content: question },
        ],
    });

    return resp.choices[0].message?.content;
    });

    (async () => {
    console.log(await rag("Where did Harrison work?"));
    })();
    ```

    </CodeGroup>

1. Call the application again to create a run:

    <CodeGroup>

    ```bash Python
    python app.py
    ```

    ```bash TypeScript
    npx ts-node app.ts
    ```

    </CodeGroup>

> 1. Return to the [platform UI](smith.langchain.com), select the **default** workspace, or the workspace you specified in [Step 2](#2-set-up-environment-variables). Select the tracing project and then the most recent run under the **Runs** tab. You'll find a trace of the entire app pipeline.

Reviewer: ```suggestion
1. Return to the [LangSmith UI](https://smith.langchain.com), navigate to the **default** Tracing Project for your workspace (or the workspace you specified in [Step 2](#2-set-up-environment-variables). You'll find a trace of the entire app pipeline.
```

<div style={{ textAlign: 'center' }}>
<img
    className="block dark:hidden"
    src="/langsmith/images/trace-quickstart-app.png"
    alt="LangSmith UI showing a trace of the entired application called rag with an input followed by an output."
/>

<img
    className="hidden dark:block"
    src="/langsmith/images/trace-quickstart-app-dark.png"
    alt="LangSmith UI showing a trace of the entired application called rag with an input followed by an output."
/>
</div>

> Tracing projects can contain a significant amount of data as you continue to trace your app. To filter traces in the UI, refer to the [Filter traces](/langsmith/filter-traces-in-application) page.

Reviewer: should we move this down to Next steps?

This will produce a trace of just the entire pipeline (with the OpenAI call as a child run) - it should look something like [this](https://smith.langchain.com/public/2174f4e9-48ab-4f9e-a8c4-470372d976f1/r)
<div style={{ textAlign: 'center' }}>
<img
    className="block dark:hidden"
    src="/langsmith/images/trace-quickstart-app.png"
    alt="LangSmith UI showing a trace of the entired application called rag with an input followed by an output."
/>

![](/langsmith/images/tracing-tutorial-chain.png)
<img
    className="hidden dark:block"
    src="/langsmith/images/trace-quickstart-app-dark.png"
    alt="LangSmith UI showing a trace of the entired application called rag with an input followed by an output."
/>
</div>

## Next steps
Tracing projects can contain a significant amount of data as you continue to trace your app. To filter traces in the UI, refer to the [Filter traces](/langsmith/filter-traces-in-application) page.

Congratulations! If you've made it this far, you're well on your way to being an expert in observability with LangSmith. Here are some topics you might want to explore next:
## Next steps

* [Tracing Integrations](/langsmith/trace-with-langchain)

> * [Send traces to a specific project](/langsmith/log-traces-to-project)

Reviewer: i would keep "Send traces to a specific project"
